phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ExpressionIndexIT.java

```diff
@@ -0,0 +1,632 @@
+		stmt.setString(7, "a.varchar" + String.valueOf(i));
+		stmt.setString(8, "a.char" + String.valueOf(i));
+		stmt.setInt(9, i);
+		stmt.setLong(10, i);
+		stmt.setBigDecimal(11, new BigDecimal((double) i));
+		stmt.setDate(12, date);
+		stmt.setString(13, "b.varchar" + String.valueOf(i));
+		stmt.setString(14, "b.char" + String.valueOf(i));
+		stmt.setInt(15, i);
+		stmt.setLong(16, i);
+		stmt.setBigDecimal(17, new BigDecimal((double) i));
+		stmt.setDate(18, date);
+		stmt.executeUpdate();
+	}
+
+	private void verifyResult(ResultSet rs, int i) throws SQLException {
+		assertTrue(rs.next());
+		assertEquals(
+				"VARCHAR"+ String.valueOf(i)+ "_"
+				+ StringUtils.rightPad("CHAR" + String.valueOf(i), 6,' ')
+				+ "_A.VARCHAR"+ String.valueOf(i)
+				+ "_"+ StringUtils.rightPad("B.CHAR" + String.valueOf(i),10, ' '), rs.getString(1));
+		assertEquals(i * 4, rs.getInt(2));
+		Date date = new Date(DateUtil.parseDate("2015-01-01 00:00:00").getTime()+(i)*NUM_MILLIS_IN_DAY);
+		assertEquals(date, rs.getDate(3));
+		assertEquals(date, rs.getDate(4));
+		assertEquals(date, rs.getDate(5));
+		assertEquals("varchar" + String.valueOf(i), rs.getString(6));
+		assertEquals("char" + String.valueOf(i), rs.getString(7));
+		assertEquals(i, rs.getInt(8));
+		assertEquals(i, rs.getLong(9));
+		assertEquals(i, rs.getDouble(10), 0.000001);
+		assertEquals(i, rs.getLong(11));
+		assertEquals(i, rs.getLong(12));
+	}
+
+	protected void helpTestCreateAndUpdate(boolean mutable, boolean localIndex)
+			throws Exception {
+		String dataTableName = mutable ? MUTABLE_INDEX_DATA_TABLE : INDEX_DATA_TABLE;
+		String fullDataTableName = INDEX_DATA_SCHEMA + QueryConstants.NAME_SEPARATOR + dataTableName;
+		Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+		Connection conn = DriverManager.getConnection(getUrl(), props);
+		try 
+		{
+			conn.setAutoCommit(false);
+			populateDataTable(conn, dataTableName);
+	
+			// create an expression index
+			String ddl = "CREATE " + (localIndex ? "LOCAL" : "") + " INDEX IDX ON " + fullDataTableName 
+			        + " ((UPPER(varchar_pk) || '_' || UPPER(char_pk) || '_' || UPPER(varchar_col1) || '_' || UPPER(char_col2))," 
+			        + " (decimal_pk+int_pk+decimal_col2+int_col1),"
+			        + " date_pk+1, date1+1, date2+1 )"
+			        + " INCLUDE (long_col1, long_col2)";
+			PreparedStatement stmt = conn.prepareStatement(ddl);
+			stmt.execute();
+	
+			// run select query with expression in WHERE clause
+			String whereSql = "SELECT long_col1, long_col2 from " + fullDataTableName + 
+			        " WHERE UPPER(varchar_pk) || '_' || UPPER(char_pk) || '_' || UPPER(varchar_col1) || '_' || UPPER(char_col2) = ?"
+			        +" AND decimal_pk+int_pk+decimal_col2+int_col1=?"
+			        // since a.date1 and b.date2 are NULLABLE and date is fixed width, these expressions are stored as DECIMAL in the index (which is not fixed width)
+			        +" AND date_pk+1=? AND cast (date1+1 as DECIMAL) =? AND cast (date2+1 as DECIMAL)=?";
```

Comment:
Actually by re-writing the functional index expressions before the visitor traversal by ExpressionCompiler will simplify things quite a bit. The tricky bit is managing what happens when a column cannot be resolved. If there are functional indexes, you don't want to stop processing, as you may find a functional index that matches still. However, if you don't find a functional index, you need to throw the ColumnNotFoundException (as local indexes tolerate this and collect them, while global ones don't). A slight tweak of this may be in order for local indexes to just collect columns that cannot be resolved (instead of catching the ColumnNotFoundException).